pageserver/src/walingest.rs

```diff
@@ -1228,6 +1238,43 @@ impl WalIngest {
         let mut rpageno = pageno % pg_constants::SLRU_PAGES_PER_SEGMENT;
         let mut page_xids: Vec<TransactionId> = vec![parsed.xid];
 
+        let now = SystemTime::now();
+        let mut warn_ingest_delay = |xact_time| {
+            match try_from_pg_timestamp(xact_time) {
+                Ok(ts) => {
+                    match now.duration_since(ts) {
+                        Ok(lag) => {
+                            if lag > modification.tline.conf.wait_lsn_timeout {
+                                self.warn_lag_ratelimit.call2(|rate_limit_stats| {
```

Comment:
I think this is already in a span that includes tenant/timeline ID, but might be worth adding a debug assertion of that.